src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,900 @@
+}
+
+CHIP_ERROR TimeSynchronizationServer::GetLocalTime(EndpointId ep, uint64_t & localTime)
+{
+    uint64_t timeZoneOffset = 0, dstOffset = 0;
+    System::Clock::Microseconds64 utcTime;
+    uint64_t chipEpochTime;
+    ReturnErrorOnFailure(System::SystemClock().GetClock_RealTime(utcTime));
+    VerifyOrReturnError(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), CHIP_ERROR_INVALID_TIME);
+    if (GetUpdatedTimeZoneState())
+    {
+        sendTimeZoneStatusEvent(ep);
+    }
+    VerifyOrReturnError(DSTState::kInvalid != GetUpdatedDSTOffsetState(), CHIP_ERROR_INVALID_TIME);
+    const auto & tz  = TimeSynchronizationServer::Instance().GetTimeZone()[0];
+    timeZoneOffset   = chip::kMicrosecondsPerSecond * static_cast<uint64_t>(tz.offset);
+    const auto & dst = TimeSynchronizationServer::Instance().GetDSTOffset()[0];
+    dstOffset        = chip::kMicrosecondsPerSecond * static_cast<uint64_t>(dst.offset);
+
+    localTime = chipEpochTime + timeZoneOffset + dstOffset;
+    return CHIP_NO_ERROR;
+}
+
+bool TimeSynchronizationServer::GetUpdatedTimeZoneState()
+{
+    System::Clock::Microseconds64 utcTime;
+    auto tzList           = TimeSynchronizationServer::Instance().GetTimeZone();
+    uint8_t activeTzIndex = 0;
+    uint64_t chipEpochTime;
+
+    VerifyOrReturnValue(System::SystemClock().GetClock_RealTime(utcTime) == CHIP_NO_ERROR, false);
+    VerifyOrReturnValue(tzList.size() != 0, false);
+    VerifyOrReturnValue(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), false);
+
+    for (uint8_t i = 0; i < tzList.size(); i++)
+    {
+        if (tzList[i].validAt != 0 && tzList[i].validAt <= chipEpochTime)
+        {
+            tzList[i].validAt = 0;
+            activeTzIndex     = i;
+        }
+    }
+    if (activeTzIndex != 0)
+    {
+        mTimeZoneListSize    = static_cast<uint8_t>(tzList.size() - activeTzIndex);
+        auto newTimeZoneList = tzList.SubSpan(activeTzIndex);
+        VerifyOrReturnValue(mTimeSyncDataProvider.StoreTimeZone(newTimeZoneList) == CHIP_NO_ERROR, false);
+        VerifyOrReturnValue(mTimeSyncDataProvider.LoadTimeZone(TimeSynchronizationServer::Instance().GetTimeZone(),
+                                                               mTimeZoneListSize) == CHIP_NO_ERROR,
+                            false);
+    }
+    return true;
+}
+
+DSTState TimeSynchronizationServer::GetUpdatedDSTOffsetState()
+{
+    System::Clock::Microseconds64 utcTime;
+    uint8_t activeDstIndex = 0;
+    auto dstList           = TimeSynchronizationServer::Instance().GetDSTOffset();
+    uint64_t chipEpochTime;
+
+    VerifyOrReturnValue(System::SystemClock().GetClock_RealTime(utcTime) == CHIP_NO_ERROR, DSTState::kInvalid);
+    VerifyOrReturnValue(dstList.size() != 0, DSTState::kInvalid);
+    VerifyOrReturnValue(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), DSTState::kInvalid);
+
+    for (uint8_t i = 0; i < dstList.size(); i++)
+    {
+        if (dstList[i].validStarting <= chipEpochTime)
```

Comment:
Ok, this case is not invalid but just means no DST is in use right now. Again, this doesn't mean the list should be emptied.

Reply:
I'm not sure I follow.  We remove the "expired" entries from the list ... unless they have all expired, in which case we leave them.  At least that's my understanding of the logic here.  Am I mis-understanding the logic?

Reply:
That's right. However, once set with valid items this list never becomes empty. From my understanding of the spec, since the last DST item has a null validUntil, it won't be removed.

Reply:
> since the last DST item has a null validUntil

Nothing requires this.  You can in fact have all the items expire.

That said, I just looked at the code again, and the issue here is not expiration; I had just mis-remembered (and possibly mis-read the code to start with?).

So the issue is: if all the DST offsets in the list _start in the future_, so that none apply now, what should happen?  The code as currently implemented is going to return `DSTState::kActive` as far as I can tell.  Is that expected in this situation?